core/controllers/dashboard.py

```diff
@@ -257,6 +258,204 @@ def _round_average_ratings(rating):
         self.render_json(self.values)
 
 
+class LearnerDashboardPage(base.BaseHandler):
+    """Page showing the user's learner dashboard."""
+
+    @base.require_user
+    def get(self):
+        if self.username in config_domain.BANNED_USERNAMES.value:
+            raise self.UnauthorizedUserException(
+                'You do not have the credentials to access this page.')
+        elif user_services.has_fully_registered(self.user_id):
+            self.values.update({
+                'nav_mode': feconf.NAV_MODE_LEARNER_DASHBOARD
+            })
+            self.render_template(
+                'pages/learner_dashboard/learner_dashboard.html',
+                redirect_url_on_logout='/')
+        else:
+            self.redirect(utils.set_url_query_parameter(
+                feconf.SIGNUP_URL, 'return_url', feconf.LEARNER_DASHBOARD_URL))
+
+
+class LearnerDashboardHandler(base.BaseHandler):
+    """Provides data for the user's learner dashboard page."""
+
+    GET_HANDLER_ERROR_RETURN_TYPE = feconf.HANDLER_TYPE_JSON
+
+    def get(self):
+        """Handles GET requests."""
+        if self.user_id is None:
+            raise self.PageNotFoundException
+
+        incomplete_exploration_summaries = []
+        incomplete_exploration_ids = (
+            learner_progress_services.get_all_incomplete_exp_ids(self.user_id))
+
+        index = 0
+        number_of_deleted_incomplete_explorations = 0
+        for summary in exp_services.get_exploration_summaries_matching_ids(
+                incomplete_exploration_ids):
+            if summary is not None:
+                incomplete_exploration_summaries.append(summary)
+            else:
+                number_of_deleted_incomplete_explorations += 1
+                learner_progress_services.remove_exp_from_incomplete_list(
+                    self.user_id, incomplete_exploration_ids[index])
+            index = index + 1
+
+        completed_exploration_summaries = []
+        completed_exploration_ids = (
+            learner_progress_services.get_all_completed_exp_ids(self.user_id))
+
+        index = 0
+        number_of_deleted_completed_explorations = 0
+        for summary in exp_services.get_exploration_summaries_matching_ids(
```

Comment:
A lot of this feels somewhat copy-paste -- and it's also way too much logic for the controller layer. Is it possible to abstract some of the commonality out a bit, and also move these to functions in the services layer? The controller layer stuff should basically be: get data from source X, get data from source Y, get data from source Z, assemble into dict, send to frontend (and not much more than that).